### PR TITLE
[datetime] remove moment dependency

### LINKED ISSUE #1930: [blueprintjs/datetime] allow for optional moment locales
react-day-picker is trying to get `.firstDayOfWeek()` but its trying to get its through `moment.localeData`
https://github.com/gpbl/react-day-picker/blob/ac85a53929e47ba025b523b05d3f6e1c7f112da7/src/addons/MomentLocaleUtils.js#L3

which is commonly excluded in webpack via:

```
const plugins = [
    new webpack.IgnorePlugin(/^\.\/locale$/, /moment$/),
];
```

this will break for some i18n users

quick solution is to add back locales, but we remove it since it increases build size. it would be nice to be able to use the DatePicker without include locales

### PR DESCRIPTION
**Pull request by giladgray**
#### Fixes #1930

⚠️  This PR is still a work-in-progress. Expecting test failures, but I'd like to start getting feedback on API changes.

#### Checklist

- [x] Fix `DateInput` tests
- [x] Fix `DateRangeInput` tests
- [x] Add tests for new methods
- [x] Update documentation
- [x] ~~`format.placeholder` or~~ top-level `placeholder` prop, etc.
- [x] ~~can I remove any component state fields?~~ if it ain't broke
  - lots of strings stored in state, but we can probably just `formatDate()` at render time

#### Changes proposed in this pull request:

- new `IDateFormatProps` interface extended by `DateInput` and `DateRangeInput`
  - 🔥  new required `formatDate` and `parseDate` callbacks
  - 🔥  `format` prop is now just an optional string that is passed to callbacks above
  - 🔥  `placeholder` prop is now on the component itself (nee `format.placeholder`)
- remove `moment` dependency from datetime package
  - docs-app still brings it in and uses it in examples
- delete all moment-based DateUtils
- refactor DateInput and DateRangeInput to use `IDateFormatter` instead of moment

#### Reviewers should focus on:

- functionality parity with http://blueprintjs.com/docs/v2/#datetime/dateinput and http://blueprintjs.com/docs/v2/#datetime/daterangeinput

### CODE REVIEW

**Comment by giladgray** [comment]
@adidahiya thoughts on borrowing r-d-p's `DayPickerInput` API? http://react-day-picker.js.org/api/DayPickerInput#formatDate

`format?: string`, `formatDate: fn`, `parseDate: fn` as top-level props, instead of object.

File: packages/core/src/components/forms/radioGroup.tsx [hunk 1]
```diff
@@ -49,7 +49,7 @@ export interface IRadioGroupProps extends IProps {
     options?: IOptionProps[];
 
     /** Value of the selected radio. The child with this value will be `:checked`. */
-    selectedValue?: string;
+    selectedValue?: string | number;
```
Comment (adidahiya) [inline packages/core/src/components/forms/radioGroup.tsx]: remove unrelated change

Comment (giladgray) [reply packages/core/src/components/forms/radioGroup.tsx]: oh but it is related, or i wouldn't have pushed it. for `FormatSelect`, which uses the index in an array. `Radio` supports number values, so `RadioGroup` should do.

File: packages/datetime/test/dateInputTests.tsx [hunk 2]
```diff
@@ -11,24 +11,31 @@ import * as sinon from "sinon";
...
 import { Months } from "../src/common/months";
+import { IDateFormatter } from "../src/dateFormatter";
 import { Classes, DateInput, TimePicker, TimePickerPrecision } from "../src/index";
 import * as DateTestUtils from "./common/dateTestUtils";
 
+const format: IDateFormatter = {
+    dateToString: date => date.toLocaleString(),
+    placeholder: "enter date",
+    stringToDate: str => new Date(Date.parse(str)),
+};
```
Comment (giladgray) [inline packages/datetime/test/dateInputTests.tsx]: big problems with this formatter. `Date.parse(str)` uses UTC timezone for "YYYY-MM-DD" strings, so some dates come out a day earlier than expected. should I just use a moment formatter in these tests?

Comment (giladgray) [reply packages/datetime/test/dateInputTests.tsx]:
```js
new Date("2/1/2015")
> "Sun Feb 01 2015 00:00:00 GMT-0800 (PST)"

new Date("2015-02-01")
> "Sat Jan 31 2015 16:00:00 GMT-0800 (PST)"
```

File: packages/core/src/components/forms/radioGroup.tsx (hunk 1 above)
Comment (adidahiya) [reply packages/core/src/components/forms/radioGroup.tsx]: separate PR please

File: packages/datetime/src/dateInput.tsx [hunk 3]
```diff
@@ -165,24 +148,17 @@ export class DateInput extends AbstractPureComponent<IDateInputProps, IDateInput
...
-        const defaultValue = this.props.defaultValue ? fromDateToMoment(this.props.defaultValue) : moment(null);
-
-        this.state = {
-            isInputFocused: false,
-            isOpen: false,
-            value: this.props.value !== undefined ? fromDateToMoment(this.props.value) : defaultValue,
-            valueString: null,
-        };
-    }
+    public state: IDateInputState = {
```
Comment (adidahiya) [inline packages/datetime/src/dateInput.tsx]: are you sure it's ok to move this to the static member? why did we have it in the constructor before?

File: packages/datetime/src/dateInput.tsx [hunk 4]
```diff
@@ -300,43 +263,33 @@ export class DateInput extends AbstractPureComponent<IDateInputProps, IDateInput
...
+    private hasMonthChanged(prevDate: Date | null, nextDate: Date | null) {
+        return (prevDate == null) !== (nextDate == null) || nextDate.getMonth() !== prevDate.getMonth();
     }
 
-    private hasTimeChanged(prevMomentDate: moment.Moment, nextMomentDate: moment.Moment) {
+    private hasTimeChanged(prevDate: Date | null, nextDate: Date | null) {
+        if (this.props.timePrecision == null) {
+            return false;
+        }
+        // TODO: getTime()?
```
Comment (adidahiya) [inline packages/datetime/src/dateInput.tsx]: - [ ] resolve this TODO in this PR

File: packages/datetime/src/datePickerCore.tsx [hunk 5]
```diff
@@ -17,14 +17,40 @@ export interface IDatePickerModifiers {
...
+     * The special value `null` indicates the absence of a date.
+     */
+    formatDate(date: Date | null, format?: string, locale?: string): string;
+
+    /**
+     * Function to deserialize a string to a JavaScript `Date` object.
+     * Return `false` if the string is an invalid date.
+     * Return `null` to represent the absence of a date.
+     */
+    parseDate(str: string, format?: string, locale?: string): Date | false | null;
```
Comment (adidahiya) [inline packages/datetime/src/datePickerCore.tsx]: not a huge blocker, but curious about this API decision: why would a consumer expect `format` in this function? if they're providing `format` as a prop, they already know the format they're interested in. in the usage of `props.parseDate()` it seems like you're passing `props.format` directly anyway.

File: packages/datetime/test/isotest.js [hunk 6]
```diff
@@ -8,9 +8,18 @@ const React = require("react");
 const DateTime = require("../lib/cjs");
 
 describe("DateTime isomorphic rendering", () => {
+    const format = {
```
Comment (adidahiya) [inline packages/datetime/test/isotest.js]: nit: `formatProps`

**Comment by adidahiya** [comment]
I didn't look at the logic very closely but I assume it's covered by the extensive test suite

File: packages/datetime/src/datePickerCore.tsx (hunk 5 above)
Comment (giladgray) [reply packages/datetime/src/datePickerCore.tsx]: prior art: http://react-day-picker.js.org/api/DayPickerInput#formatDate.

this is the only place the `format` prop gets used, so we could drop it from the API and push it to the user. but the nice thing about this usage is that only one set of moment utils for `format/parseDate` could be used for all formats.

File: packages/datetime/src/dateInput.tsx (hunk 3 above)
Comment (giladgray) [reply packages/datetime/src/dateInput.tsx]: i am quite sure. typescript always inline state init in the constructor that it generates (before all the other `=>` bound methods), so you can actually safely use `this.props` in `public state` initialization.

File: packages/datetime/src/dateFormat.tsx [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+ */
+
+import { isDateValid, isDayInRange } from "./common/dateUtils";
+import { IDatePickerBaseProps } from "./datePickerCore";
+
+export interface IDateFormatProps {
+    /**
+     * Date format string, passed to `formatDate` and `parseDate`.
+     */
+    format?: string;
```
Comment (giladgray) [inline packages/datetime/src/dateFormat.tsx]: @adidahiya what's your take on this prop? worth keeping?

Comment (adidahiya) [reply packages/datetime/src/dateFormat.tsx]: prefer to keep it simple and add it later if it's desired

**Comment by adidahiya** [comment]
code looks fine but this radio doesn't work for me:

image

**Comment by giladgray** [comment]
@adidahiya not sure what you mean with format screenshot, works for me. i just removed the (def broken) "from now" option.

**Comment by adidahiya** [comment]
we should probably still add a `@blueprintjs/datetime-moment` package after this PR
